File: packages/core/src/policy/outcome.ts
/** @fileoverview Delivery outcome classification (§8, G9). @module @anyhook/core */
import type { AttemptStatus } from '../types/attempt.js';

export type Outcome = 'delivered' | 'retryable' | 'permanent';

/**
 * Classify a delivery attempt's transport status (§8):
 * - 2xx                → `delivered`
 * - 3xx                → `permanent` (redirects are not followed; a redirect on a webhook target is misconfiguration)
 * - 429                → `retryable`
 * - other 4xx          → `permanent` (the receiver rejected it; retrying won't help)
 * - 5xx                → `retryable`
 * - timeout / network  → `retryable`
 * - <200 (unexpected)  → `retryable`
 */
export function classifyOutcome(status: AttemptStatus): Outcome {
  if (status === 'timeout' || status === 'network') return 'retryable';
  if (status >= 200 && status < 300) return 'delivered';
  if (status === 429) return 'retryable';
  if (status >= 300 && status < 500) return 'permanent';
  if (status >= 500) return 'retryable';
  return 'retryable'; // 1xx / anomalous
}

File: go/core/outcome.go
package core

// Outcome is the classification of a single delivery attempt's transport status.
type Outcome string

const (
	OutcomeDelivered Outcome = "delivered"
	OutcomeRetryable Outcome = "retryable"
	OutcomePermanent Outcome = "permanent"
)

// ClassifyOutcome classifies a delivery attempt's transport status:
//
//   - 2xx                -> delivered
//   - 3xx                -> permanent (redirects are not followed; a redirect
//     on a webhook target is misconfiguration)
//   - 429                -> retryable
//   - other 4xx          -> permanent (the receiver rejected it; retrying won't help)
//   - 5xx                -> retryable
//   - timeout / network  -> retryable
//   - <200 (unexpected)  -> retryable
func ClassifyOutcome(status AttemptStatus) Outcome {
	if status == StatusTimeout || status == StatusNetwork {
		return OutcomeRetryable
	}
	s := int(status)
	if s >= 200 && s < 300 {
		return OutcomeDelivered
	}
	if s == 429 {
		return OutcomeRetryable
	}
	if s >= 300 && s < 500 {
		return OutcomePermanent
	}
	if s >= 500 {
		return OutcomeRetryable
	}
	return OutcomeRetryable // 1xx / anomalous
}
